### PR TITLE
Travis: run full test suite on native macOS 

### PR DESCRIPTION
**Pull request by Sjors**
Adds an additional Travis machine to run the functional test suite on native macOS 

Homebrew is not particularly Travis compatible, but I found some useful hints here: https://discourse.brew.sh/t/best-practice-for-homebrew-on-travis-brew-update-is-5min-to-build-time/5215/11

### CODE REVIEW

**Comment by emilengler** [comment]
Are there travis tests which can proof it? Like your fork?
But I generally like this idea as the current situation with tuxes on the side is a bit confusing

File: .travis.yml [hunk 1]
```diff
@@ -225,3 +225,25 @@ jobs:
...
+    os: osx
+    # Use the most recent version:
+    # Xcode 11, macOS 10.14, JDK 12.0.1
+    # https://docs.travis-ci.com/user/reference/osx/#OS-X-Version
+    osx_image: xcode11
+    language:
+      - ruby
+      - python
+    python:
+        - "3.6.5"
```
Comment (fanquake) [inline .travis.yml]: Should this be `3.5.6` to match our [minimum supported Python](https://github.com/bitcoin/bitcoin/blob/master/.python-version)?

**Comment by fanquake** [comment]
Concept ACK. Will take a look at the base PR first.

**Comment by Sjors** [comment]
Travis is ignoring this PR and my branch atm.

**Comment by emilengler** [comment]
Can you enable travis for your forked repository?

**Comment by Sjors** [comment]
@emilengler I tried that, but it wasn't building there either. I'll wait until the previous PR is merged and then try again.

**Comment by emilengler** [comment]
@Sjors Ok I will fork your repository, then remove it from the parent and try to enable travis then

**Comment by emilengler** [comment]
@Sjors Doesn't work, take a look [https://travis-ci.org/emilengler/bitcoin-sjors-test/requests](here).
I also forked your [repo](https://github.com/emilengler/bitcoin-sjors-test) here as a playgroud, I also invited you for commit access, will delete it once it is done.

File: ci/test/00_setup_env_mac.sh [hunk 2]
```diff
@@ -7,7 +7,7 @@
 export LC_ALL=C.UTF-8
 
 export HOST=x86_64-apple-darwin14
-export PACKAGES="cmake imagemagick libcap-dev librsvg2-bin libz-dev libbz2-dev libtiff-tools python3-dev python3-setuptools"
+export BREW_PACKAGES="automake berkeley-db4 libtool boost miniupnpc openssl pkg-config protobuf qt qrencode librsvg pyenv"
```
Comment (MarcoFalke) [inline ci/test/00_setup_env_mac.sh]: why is this changed?

File: ci/test/04_install.sh [hunk 3]
```diff
@@ -4,6 +4,11 @@
 # Distributed under the MIT software license, see the accompanying
 # file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+if [ "$TRAVIS_OS_NAME" == "osx" ]; then
+    travis_retry brew install $BREW_PACKAGES
```
Comment (MarcoFalke) [inline ci/test/04_install.sh]:
```suggestion
    ${CI_RETRY_EXE} brew install $PACKAGES
```

**Review by maflcko — COMMENTED** [review]
Concept ACK

File: ci/test/00_setup_env_mac.sh (hunk 2 above)
Comment (Sjors) [reply ci/test/00_setup_env_mac.sh]: By accident. Fixed. This PR is still very rough; I can pick it up again now that #13728 is merge.

File: ci/test/04_install.sh [hunk 4]
```diff
@@ -4,6 +4,11 @@
 # Distributed under the MIT software license, see the accompanying
 # file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+if [ "$TRAVIS_OS_NAME" == "osx" ]; then
+    ${CI_RETRY_EXE} brew install $BREW_PACKAGES
+    return 0
+ fi
+
```
Comment (MarcoFalke) [inline ci/test/04_install.sh]: should move this below the LC_ALL line

File: .travis.yml [hunk 5]
```diff
@@ -160,3 +160,20 @@ jobs:
       name: 'macOS 10.10  [GOAL: deploy] [no functional tests]'
       env: >-
         FILE_ENV="./ci/test/00_setup_env_mac.sh"
+
+    - stage: test
+      name: 'macOS 10.14 native'
+      os: osx
+      # Use the most recent version:
+      # Xcode 11, macOS 10.14, JDK 12.0.1
+      # https://docs.travis-ci.com/user/reference/osx/#OS-X-Version
```
Comment (MarcoFalke) [inline .travis.yml]:
```suggestion
      # https://docs.travis-ci.com/user/reference/osx/#macos-version
```

File: .travis.yml [hunk 6]
```diff
@@ -160,3 +160,20 @@ jobs:
...
+      osx_image: xcode11
+      language:
+        - ruby
+        - python
+      python:
+        - "3.5.6"
+      rvm:
+        - 2.6.0
```
Comment (MarcoFalke) [inline .travis.yml]: This config doesn't boot (even after waiting an hour)

Maybe you need to install python via `brew upgrade python`?

Comment (Sjors) [reply .travis.yml]: Trying with `pyenv install 3.5.6`

Comment (Sjors) [reply .travis.yml]: That doesn't seem to boot either.

Comment (MarcoFalke) [reply .travis.yml]: In def61133e18a4487af1c9406ffa3f6bf19b702fd I removed both of these and it fails with:

https://travis-ci.org/bitcoin/bitcoin/jobs/596568166

Comment (MarcoFalke) [reply .travis.yml]:
```
Warning: automake 1.16.1_1 is already installed and up-to-date

To reinstall 1.16.1_1, run `brew reinstall automake`

Warning: berkeley-db@4 4.8.30 is already installed and up-to-date

To reinstall 4.8.30, run `brew reinstall berkeley-db@4`

Warning: libtool 2.4.6_1 is already installed and up-to-date

To reinstall 2.4.6_1, run `brew reinstall libtool`

Warning: boost 1.71.0 is already installed and up-to-date

To reinstall 1.71.0, run `brew reinstall boost`

Warning: miniupnpc 2.1 is already installed and up-to-date

To reinstall 2.1, run `brew reinstall miniupnpc`

Warning: openssl 1.0.2t is already installed and up-to-date

To reinstall 1.0.2t, run `brew reinstall openssl`

Warning: pkg-config 0.29.2 is already installed and up-to-date

To reinstall 0.29.2, run `brew reinstall pkg-config`

Warning: qt 5.13.1 is already installed and up-to-date

To reinstall 5.13.1, run `brew reinstall qt`

Warning: qrencode 4.0.2 is already installed and up-to-date

To reinstall 4.0.2, run `brew reinstall qrencode`

Warning: librsvg 2.46.1 is already installed and up-to-date

To reinstall 2.46.1, run `brew reinstall librsvg`

Warning: pyenv 1.2.13_1 is already installed and up-to-date

To reinstall 1.2.13_1, run `brew reinstall pyenv`

Error: protobuf 3.9.1_1 is already installed

To upgrade to 3.10.0, run `brew upgrade protobuf`.

The command "brew install automake berkeley-db4 libtool boost miniupnpc openssl pkg-config protobuf qt qrencode librsvg pyenv" failed. Retrying, 3 of 3.

Comment (Sjors) [reply .travis.yml]: Oh boy, Homebrew is a pain. I lifted some stuff from:
https://discourse.brew.sh/t/best-practice-for-homebrew-on-travis-brew-update-is-5min-to-build-time/5215/11 and https://github.com/vadi2/Mudlet/blob/857d1ae29fe6e55f1507399200c45db318b28ae6/CI/travis.osx.install.sh#L11-L35

Comment (Sjors) [reply .travis.yml]: @vadi2 @keneanung let me know if you prefer not to be mentioned as co-authored-by in the commit.

Comment (Sjors) [reply .travis.yml]: (assuming we ever get this to work :-)

Comment (vadi2) [reply .travis.yml]: It's fine! Let us know if we can help you out with this.

**Comment by Sjors** [comment]
5ea77cd56d [failed](https://travis-ci.org/bitcoin/bitcoin/jobs/597230378) at `travis_fold:end:autogen` without any error. I added a check for OSX in order to run `./autogen.sh` without `DOCKER_EXEC`; who knows...

File: ci/test/00_setup_env_mac_functional.sh [hunk 7]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+#
+# Copyright (c) 2019 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+export LC_ALL=C.UTF-8
+
+export HOST=x86_64-apple-darwin14
+export BREW_PACKAGES="automake berkeley-db4 libtool boost miniupnpc openssl pkg-config protobuf qt qrencode librsvg pyenv ccache"
```
Comment (fanquake) [inline ci/test/00_setup_env_mac_functional.sh]: No reason to install `protobuf` if we're not enabling BIP70.

File: ci/test/00_setup_env_mac_functional.sh [hunk 8]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+#
+# Copyright (c) 2019 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+export LC_ALL=C.UTF-8
+
```
Comment (fanquake) [inline ci/test/00_setup_env_mac_functional.sh]: Even thought it'll be unused, could add `NO_DEPENDS=1` here to make it clear that we aren't actually using depends.

File: ci/test/00_setup_env_mac_functional.sh (hunk 7 above)
Comment (Sjors) [reply ci/test/00_setup_env_mac_functional.sh]: Dropped (in other branch, will push later)

Comment (luke-jr) [reply ci/test/00_setup_env_mac_functional.sh]: If we're building/testing only once natively, we should enable everything...

File: .travis.yml (hunk 6 above)
Comment (keneanung) [reply .travis.yml]: I'm good with this as well. As Vadim said, let us know if we can be of help.

Comment (Sjors) [reply .travis.yml]: Thanks! I could use some help with _update 7_ below.

File: ci/test/00_setup_env_mac_functional.sh (hunk 7 above)
Comment (Sjors) [reply ci/test/00_setup_env_mac_functional.sh]: I added it back and enabled BIP70 in the build.

**Comment by Sjors** [comment]
Travis macOS passed! Hopefully I didn't jinx it by adding `pip install zmq` for the functional ZMQ tests...

File: ci/test/04_install.sh [hunk 9]
```diff
@@ -26,7 +73,7 @@ elif [[ $BITCOIN_CONFIG = *--with-sanitizers=*address* ]]; then # If ran with (A
   DOCKER_ADMIN="--cap-add SYS_PTRACE"
 fi
 
-if [ -z "$RUN_CI_ON_HOST" ]; then
+if [ -z "$RUN_CI_ON_HOST" ] && [ "$TRAVIS_OS_NAME" != "osx" ]; then
```
Comment (MarcoFalke) [inline ci/test/04_install.sh]: I'd prefer if you set `RUN_CI_ON_HOST=1` in the env file or travis.yml

File: ci/test/04_install.sh [hunk 10]
```diff
@@ -6,6 +6,53 @@
...
+      echo "Attempt ${RETRIES} failed."
+      if [ "${RETRIES}" -eq 3 ]; then
+        echo "Too many retries. Aborting."
+        exit 1
+      else
+        echo "Retrying..."
+      fi
+    done
+
+  done
```
Comment (MarcoFalke) [inline ci/test/04_install.sh]: homebrew :man_facepalming: 


This can't be written shorter, right?

**Review by maflcko — COMMENTED** [review]
ACK 26068f31ad60c957ab7e6aef0a757fea4397da13 (looked at the diff on GitHub)

File: ci/test/04_install.sh (hunk 10 above)
Comment (Sjors) [reply ci/test/04_install.sh]: Probably, but I'm no bash-master. Homebrew doesn't have a "install or update if possible, don't complain otherwise" feature, because ... reason.

Comment (fanquake) [reply ci/test/04_install.sh]: Could you just run `brew update` then `brew install package-name --force` for each package you're trying to install?

Comment (Sjors) [reply ci/test/04_install.sh]: Reinstalling these packages at every Travis run seems very slow:
```
-f, --force
              Remove a previously cached version and re-fetch.
```

Comment (MarcoFalke) [reply ci/test/04_install.sh]: I'd be fine with a solution that works, but isn't the fastest. We can optimize for speed in a follow up pull request.

File: ci/lint/04_install.sh [hunk 11]
```diff
@@ -8,6 +8,8 @@ export LC_ALL=C
 
 if [ "$TRAVIS_OS_NAME" == "osx" ]; then
   # update first to install required ruby dependency
+  cd /usr/local/Homebrew
+  git reset --hard origin/master
```
Comment (MarcoFalke) [inline ci/lint/04_install.sh]: Why is the linter modified?

Comment (Sjors) [reply ci/lint/04_install.sh]: Oops, wrong one.

File: ci/test/04_install.sh [hunk 12]
```diff
@@ -6,6 +6,29 @@
 
 export LC_ALL=C.UTF-8
 
+if [ "$TRAVIS_OS_NAME" == "osx" ]; then
+  cd /usr/local/Homebrew || exit 1
+  git reset --hard origin/master
+  ${CI_RETRY_EXE} brew update
```
Comment (Sjors) [inline ci/test/04_install.sh]: @MarcoFalke `brew update` blew up on a previous run (can't find it anymore) and recommended these two lines to "fix" it.

Comment (Sjors) [reply ci/test/04_install.sh]: It might be worth rebasing the PR for a while to see if any other surprises show up.

File: ci/test/06_script_b.sh [hunk 13]
```diff
@@ -6,7 +6,10 @@
 
 export LC_ALL=C.UTF-8
 
+# Temporarily disable errexit, because Travis macOS fails without error message
+set +o errexit
```
Comment (Sjors) [inline ci/test/06_script_b.sh]: This is where your sanity gives in...

File: .travis.yml (hunk 6 above)
Comment (keneanung) [reply .travis.yml]: Hm. I'm unsure whether we had this error... And since we don't use this deep caching anymore, we don't have any further experience in this dark corner of homebrew :ghost:

**Comment by MarcoFalke** [comment]
re-ACK 1f6c650c99eb94fdd0cbdd528aa3405201de08c2

File: ci/test/04_install.sh [hunk 14]
```diff
@@ -6,6 +6,32 @@
 
 export LC_ALL=C.UTF-8
 
+if [ "$TRAVIS_OS_NAME" == "osx" ]; then
+  set +o errexit
+  pushd /usr/local/Homebrew || exit 1
+  git reset --hard origin/master
+  popd || exit 1
+  set -o errexit
```
Comment (MarcoFalke) [inline ci/test/04_install.sh]: What happens if you remove this?

File: .travis.yml [hunk 15]
```diff
@@ -34,6 +34,13 @@ cache:
     - $TRAVIS_BUILD_DIR/depends/built
     - $TRAVIS_BUILD_DIR/depends/sdk-sources
     - $TRAVIS_BUILD_DIR/ci/scratch/.ccache
+    # macOS
+    - $HOME/Library/Caches/Homebrew
+    - /usr/local/Homebrew
+before_cache:
+  - if [ "${TRAVIS_OS_NAME}" = "osx" ]; then brew cleanup; fi
+  # Cache only .git files under "/usr/local/Homebrew" so "brew update" does not take 5min every build
+  # - if [ "${TRAVIS_OS_NAME}" = "osx" ]; then find /usr/local/Homebrew \! -regex ".+\.git.+" -delete; fi
```
Comment (MarcoFalke) [inline .travis.yml]: Is this still needed?

The whole process takes 2 min without it: https://travis-ci.org/bitcoin/bitcoin/jobs/597848502#L100

File: ci/test/05_before_script.sh [hunk 16]
```diff
@@ -6,19 +6,28 @@
...
-DOCKER_EXEC echo \> \$HOME/.bitcoin  # Make sure default datadir does not exist and is never read by creating a dummy file
+# Make sure default datadir does not exist and is never read by creating a dummy file
+if [ "$TRAVIS_OS_NAME" == "osx" ]; then
+  echo > $HOME/Library/Application\ Support/Bitcoin
+else
+  DOCKER_EXEC echo \> \$HOME/.bitcoin
+fi
 
-mkdir -p depends/SDKs depends/sdk-sources
+if [ "$TRAVIS_OS_NAME" != "osx" ]; then
```
Comment (ryanofsky) [inline ci/test/05_before_script.sh]: Maybe should add a comment explaining this check. I'm confused about why all of this is skipped on osx. Does the depends build not work there? It would seem cleaner to use NO_DEPENDS variable to skip depends steps than to tie the decision to the travis operating system name.

**Review by ryanofsky — COMMENTED** [review]
Post merge code review ACK 1f6c650c99eb94fdd0cbdd528aa3405201de08c2

Since this adds homebrew, I wonder if it fixes the scripted diff lint error I ran into https://github.com/bitcoin/bitcoin/pull/13728#issuecomment-542973824

File: ci/test/05_before_script.sh (hunk 16 above)
Comment (MarcoFalke) [reply ci/test/05_before_script.sh]: Thanks for the review @ryanofsky . This check looks indeed redundant. Removed in #17176

File: ci/test/06_script_a.sh [hunk 17]
```diff
@@ -20,7 +20,11 @@ fi
 END_FOLD
 
 mkdir -p build
+
+# Temporarily disable errexit, because Travis macOS fails without error message
+set +o errexit
 cd build || (echo "could not enter build directory"; exit 1)
+set -o errexit
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply ci/test/06_script_a.sh]: Filed as good first issue. See #17178

File: .travis.yml (hunk 15 above)
Comment (Sjors) [reply .travis.yml]: I don't know if it builds up any cruft over time, but if time doesn't creep up, then we can drop this.

Comment (MarcoFalke) [reply .travis.yml]: Well it is commented out already, so I removed it in #17176
